feat: upgrade react-native-webview

jpcloureiro opened the pull request:
## **Description**

Moving away from `.patch` changes on the webview

Our [RNWebView fork](https://github.com/MetaMask/react-native-webview-mm) contains the security patches. 
 The "Javascript Injection" patch was removed and the related patches to fix bugs introduced by that.

Jest snapshots updates to reflect the new RNWebView component

<!--
Write a short description of the changes included in this pull request, and also include relevant motivation and context. Have in mind the following questions:
1. What is the reason for the change?
2. What is the improvement/solution?
-->

## **Related issues**

Addresses: [mobile-planning #1294](https://github.com/MetaMask/mobile-planning/issues/1294)

## Acceptance Criteria
- [x] https://github.com/MetaMask/mobile-planning/issues/1618 ~must be closed with zero bugs~ Done

## **Pre-merge author checklist**

- [ ] I’ve followed [MetaMask Coding Standards](https://github.com/MetaMask/metamask-mobile/blob/main/.github/guidelines/CODING_GUIDELINES.md).
- [ ] I've clearly explained what problem this PR is solving and how it is solved.
- [ ] I've linked related issues
- [ ] I've included manual testing steps
- [ ] I've included screenshots/recordings if applicable
- [ ] I’ve included tests if applicable
- [ ] I’ve documented my code using [JSDoc](https://jsdoc.app/) format if applicable
- [ ] I’ve applied the right labels on the PR (see [labeling guidelines](https://github.com/MetaMask/metamask-mobile/blob/main/.github/guidelines/LABELING_GUIDELINES.md)). Not required for external contributors.
- [ ] I’ve properly set the pull request status:
  - [ ] In case it's not yet "ready for review", I've set it to "draft".
  - [ ] In case it's "ready for review", I've changed it from "draft" to "non-draft".

## **Pre-merge reviewer checklist**

- [ ] I've manually tested the PR (e.g. pull and build branch, run the app, test code being changed).
- [ ] I confirm that this PR addresses all acceptance criteria described in the ticket it closes and includes the necessary testing evidence such as recordings and or screenshots.

jpcloureiro committed:
Port: remove webviewref validation

jpcloureiro commented:
Failing tests might require patching `react-native-webview`

https://github.com/react-native-webview/react-native-webview/issues/2959#issuecomment-1715421671

jpcloureiro commented on app/core/BackgroundBridge/Port.ts:
```diff
@@ -19,9 +19,7 @@ class Port extends EventEmitter {
     const js = this._isMainFrame
       ? JS_POST_MESSAGE_TO_PROVIDER(msg, origin)
       : JS_IFRAME_POST_MESSAGE_TO_PROVIDER(msg, origin);
-    if (this._window.webViewRef?.current) {
-      this._window?.injectJavaScript(js);
```
This change is related to the [refactor](https://github.com/react-native-webview/react-native-webview/pull/2508) that was shipped on[ 11.22.0](https://github.com/react-native-webview/react-native-webview/releases/tag/v11.22.0)
Before 11.22.0 we had `webViewRef` inside webViewRef, that allowed the if condition to be true if webViewRef was defined.

We can now rely on `this._window` to be defined to inject JS

jpcloureiro commented on ios/Podfile.lock:
```diff
@@ -81,9 +81,8 @@ PODS:
   - lottie-react-native (5.1.5):
     - lottie-ios (~> 3.4.0)
     - React-Core
-  - MultiplatformBleAdapter (0.1.9)
```
TODO: clean podfile changes to include only `react-native-webview`ones

jpcloureiro replied:
done

jpcloureiro commented on package.json:
```diff
@@ -330,7 +330,7 @@
     "react-native-video": "5.2.1",
     "react-native-view-shot": "^3.1.2",
-    "react-native-webview": "11.13.0",
+    "react-native-webview": "git+https://github.com/jpcloureiro/react-native-webview.git#8c434cad6f0ecea9e57f9ab05393004dbd315f78",
```
TODO: create fork under metamask org

jpcloureiro replied:
done https://github.com/metamask/react-native-webview-mm

jpcloureiro commented on patches/react-native-webview+11.13.0.patch:
Followup with issues to extract the commits from this patch file and apply them into the fork if still needed

jpcloureiro replied:
done https://github.com/MetaMask/mobile-planning/issues/1294

jpcloureiro commented:
We have a working in-app browser on both Android and iOS platforms with latest (13.6.2) react-native-webview.

Had to change the Android WebView callback that injects the web3 provider. See changes [here ](https://github.com/jpcloureiro/react-native-webview/commit/8c434cad6f0ecea9e57f9ab05393004dbd315f78)
> **Warning** Merging main breaks APP <-> DAPP communication.

TODO: Identify which commit from main breaks the communication.

wachunei commented:
Hey @jpcloureiro, as our ramp integrations rely heavily on the webview package I want to let you and all the mobile team know that we will be keeping an eye on the progress of this PR so by the time this is ready to be reviewed our QA engineer Brendan (@bkirb) confirms our integrations with providers are not broken.

Please consider adding `team-ramp`, `needs-ramp-qa` labels when you think that time has come 🙏

jpcloureiro committed:
use react-native-webview fork

jpcloureiro committed:
remove webview patches

jpcloureiro committed:
update podfile lock file

jpcloureiro committed:
Merge branch 'main' into upgrade/react-native-webview

jpcloureiro commented:
TODO: Test Blockaid feature (PPOMView)

jpcloureiro committed:
Improve web3 provider injection (#8130)

jpcloureiro committed:
remove temporary ignored vuln

jpcloureiro committed:
Merge branch 'main' into upgrade/react-native-webview

jpcloureiro committed:
update PPOMView jest snapshot

jpcloureiro committed:
Merge branch 'main' into upgrade/react-native-webview

jpcloureiro committed:
REVERTME: show git info on prod builds

jpcloureiro committed:
Merge branch 'main' into upgrade/react-native-webview

jpcloureiro committed:
update Quotes component jest snapshot

jpcloureiro committed:
Merge branch 'main' into upgrade/react-native-webview

jpcloureiro committed:
update jest snapshots

jpcloureiro committed:
patch: (webview upgrade) Android filename validation when downloading from browser (#8654)

This change is part of the file react-native-webview+11.13.0.patch, a
collection of patches that we need to apply on top of
react-native-webview.

Original Pull Request:
[#6844](https://github.com/MetaMask/metamask-mobile/pull/6844)

Webview changes can be reviewed here:

[MetaMask/react-native-webview-mm#7](https://github.com/MetaMask/react-native-webview-mm/pull/7)

```
Scenario: Download file with invalid file name
  Navigate to websites using browser address bar.
    Given the app displayed the splash animation
    And I have imported my wallet
    And I tap No Thanks on the Enable security check screen
    And I tap No thanks on the onboarding welcome tutorial
    And I navigate to the browser
    And I have 1 browser tab displayed
    When I download a file with invalid filename
    Then an error message should appear
    And the file is not download
```

<!-- If applicable, add screenshots and/or recordings to visualize the
before and after of your change. -->

https://github.com/MetaMask/metamask-mobile/assets/1649425/aa7d07a8-ab95-4f36-9325-bdcd6760dcd0

https://github.com/MetaMask/metamask-mobile/assets/1649425/3e3fd1cf-5593-4afb-8045-4666cad494cd

- [ ] I’ve followed [MetaMask Coding
Standards](https://github.com/MetaMask/metamask-mobile/blob/main/.github/guidelines/CODING_GUIDELINES.md).
- [ ] I've clearly explained what problem this PR is solving and how it
is solved.
- [ ] I've linked related issues
- [ ] I've included manual testing steps
- [ ] I've included screenshots/recordings if applicable
- [ ] I’ve included tests if applicable
- [ ] I’ve documented my code using [JSDoc](https://jsdoc.app/) format
if applicable
- [ ] I’ve applied the right labels on the PR (see [labeling
guidelines](https://github.com/MetaMask/metamask-mobile/blob/main/.github/guidelines/LABELING_GUIDELINES.md)).
Not required for external contributors.
- [ ] I’ve properly set the pull request status:
  - [ ] In case it's not yet "ready for review", I've set it to "draft".
- [ ] In case it's "ready for review", I've changed it from "draft" to
"non-draft".

- [ ] I've manually tested the PR (e.g. pull and build branch, run the
app, test code being changed).
- [ ] I confirm that this PR addresses all acceptance criteria described
in the ticket it closes and includes the necessary testing evidence such
as recordings and or screenshots.

jpcloureiro committed:
Merge branch 'main' into upgrade/react-native-webview

jpcloureiro committed:
update rnwebview package reference

jpcloureiro committed:
update rnwebview package reference

jpcloureiro committed:
update rnwebview package reference

jpcloureiro committed:
update webview commit hash

jpcloureiro committed:
remove build.sh changes

jpcloureiro committed:
Merge branch 'main' into upgrade/react-native-webview

jpcloureiro committed:
remove bitrise.yml changes

jpcloureiro committed:
remove AppInformation changes

jpcloureiro committed:
BrowserTab: more context about the eventListener

Add better descriptive information on why we want to wait for DOM
to be fully loaded before executing the web3 provider
for the first time

jpcloureiro committed:
update jest snapshots

wachunei commented on app/components/UI/Ramp/Views/Quotes/Quotes.test.tsx:
```diff
@@ -44,6 +45,16 @@ function render(Component: React.ComponentType) {
 
 jest.unmock('react-redux');
 
+jest.mock('react-native-webview', () => {
+  // eslint-disable-next-line @typescript-eslint/no-require-imports, @typescript-eslint/no-var-requires
+  const { View } = require('react-native');
+  const WebView = (props: ViewProps) => <View {...props} />;
+
+  return {
+    WebView,
+  };
+});
+
```
Is there a reason to mock the package this way? now snapshots will render a `<View>` with the props instead of `<RNCWebView>`

wachunei replied:
Thank you, would it be better to have this mock in the setup file instead `app/util/test/testSetup.js`?

jpcloureiro replied:
addressed in 00b15b4

wachunei replied:
Nice, I forgot to mention it before, but for future context: can you add the link as a comment in the mock please? For anyone running into it to understand why it's there 🙏

jpcloureiro replied:
Comment added in [4b63ebf](https://github.com/MetaMask/metamask-mobile/pull/7759/commits/4b63ebf64a134c5f0dc396193b0bbc40d6063646)

jpcloureiro committed:
Merge branch 'main' into upgrade/react-native-webview

jpcloureiro committed:
format code

jpcloureiro committed:
move react-native-webview mock to testSetup

jpcloureiro committed:
update SimpleWebview jest snapshot

jpcloureiro committed:
webview jest mock: add GH issue for more context

Add note comment to provide more context on why we
need to mock webview to run jest

jpcloureiro committed:
Merge branch 'main' into upgrade/react-native-webview

jpcloureiro committed:
version bump: set version up for testflight upload

jpcloureiro commented:
Description updated with acceptance criteria definition

jpcloureiro committed:
Revert "version bump: set version up for testflight upload"

This reverts commit 1849b1a21e91a7af5c23e4bc75c336c0df50274d.

jpcloureiro committed:
Merge branch 'main' into upgrade/react-native-webview

naugtur commented on package.json:
```diff
@@ -315,7 +315,7 @@
     "react-native-video": "5.2.1",
     "react-native-view-shot": "^3.1.2",
-    "react-native-webview": "11.13.0",
+    "react-native-webview": "https://github.com/MetaMask/react-native-webview-mm#b27669c0dddf5420a3147113d475f3fbe1a27f60",
```
sidenote - have we considered contributing the changes upstream?

jpcloureiro replied:
@legobeat PR #9695 was merged into this. It has the necessary changes to use `@metamask/react-native-webview` from NPM

jpcloureiro commented:
Moving PR to Draft as there are still debug commits coming in.

I'll rebase and mark it as ready for review once we finish the debugging phase.

jpcloureiro committed:
remove event listener

jpcloureiro committed:
Merge branch 'main' into upgrade/react-native-webview

jpcloureiro committed:
conditional event listener

jpcloureiro committed:
test reloading page if documentElement is null

jpcloureiro committed:
explain Android web3 provider injection workaround

jpcloureiro committed:
Merge branch 'main' into upgrade/react-native-webview

jpcloureiro committed:
Merge branch 'main' into upgrade/react-native-webview

jpcloureiro committed:
Merge branch 'main' into upgrade/react-native-webview

jpcloureiro committed:
test webview fork last commit

tommasini committed:
Merge branch 'main' into upgrade/react-native-webview

sethkfman committed:
Merge branch 'main' into upgrade/react-native-webview

cortisiko committed:
Merge branch 'main' into upgrade/react-native-webview

tommasini committed:
chore: migrate to mm webview npm registry (#9695)

## **Description**

This PR migrates `react-native-webview` to
`@metamask/react-native-webview` on the webview upgrade branch branch

This is the [PR](https://github.com/MetaMask/metamask-mobile/pull/7759)
of the main webview upgrade branch.

* Removed react native animated fox package and created the component of
react native animated fox.

## **Related issues**

Fixes:

## **Manual testing steps**

1. Go to this page...
2.
3.

## **Screenshots/Recordings**

<!-- If applicable, add screenshots and/or recordings to visualize the
before and after of your change. -->

### **Before**

<!-- [screenshots/recordings] -->

### **After**

<!-- [screenshots/recordings] -->

## **Pre-merge author checklist**

- [ ] I’ve followed [MetaMask Coding
Standards](https://github.com/MetaMask/metamask-mobile/blob/main/.github/guidelines/CODING_GUIDELINES.md).
- [ ] I've completed the PR template to the best of my ability
- [ ] I’ve included tests if applicable
- [ ] I’ve documented my code using [JSDoc](https://jsdoc.app/) format
if applicable
- [ ] I’ve applied the right labels on the PR (see [labeling
guidelines](https://github.com/MetaMask/metamask-mobile/blob/main/.github/guidelines/LABELING_GUIDELINES.md)).
Not required for external contributors.

## **Pre-merge reviewer checklist**

- [ ] I've manually tested the PR (e.g. pull and build branch, run the
app, test code being changed).
- [ ] I confirm that this PR addresses all acceptance criteria described
in the ticket it closes and includes the necessary testing evidence such
as recordings and or screenshots.

---------

Co-authored-by: legobt <6wbvkn0j@anonaddy.me>
Co-authored-by: sethkfman <10342624+sethkfman@users.noreply.github.com>
Co-authored-by: Curtis <Curtis.David7@gmail.com>

jpcloureiro committed:
Merge branch 'main' into upgrade/react-native-webview

jpcloureiro commented:
Uniswap fails to initialize web3 provider
![Screenshot 2024-06-04 at 12 55 17](https://github.com/MetaMask/metamask-mobile/assets/1649425/0ee6dc06-d4c0-4a89-b188-8ee35a162e61)

Error logged on dev-console
![Screenshot 2024-06-04 at 12 54 25](https://github.com/MetaMask/metamask-mobile/assets/1649425/5a1bcc03-c4ea-4644-aa17-f91c94e90058)


code tested on commit 268d8d87c55d6d9b30c74083ee92a0661558292d

jpcloureiro commented:
> @jpcloureiro Is this still being pursued independently from #9695 and how are they different?

@legobeat #9695 was merged into this. We'll continue the webview work on this PR.

jpcloureiro committed:
update podfile lock

legobeat committed:
Merge branch 'main' into upgrade/react-native-webview

jpcloureiro committed:
RNWebView upgrade: Fix e2e tests on CI (#9935)

Update detox matchers to grab the correct webview instance to test.

Even though e2e tests pass locally, they fail on the CI

Using a combination of detox matchers, such as, by.id, by.type and
withAncestor. we can select the single webview instance we want
to interact with.

See also detox documentation:
https://wix.github.io/Detox/docs/guide/testing-webviews/#step-1-locating-the-webview

legobeat reviewed (dismissed):
@jpcloureiro CI does not seem to be running at all - maybe it will trigger if you sync with `main` and resolve the conflict?

jpcloureiro committed:
Merge branch 'main' into upgrade/react-native-webview

jpcloureiro committed:
update PPOMView jest snapshot

jpcloureiro committed:
Merge branch 'main' into upgrade/react-native-webview

jpcloureiro committed:
Merge branch 'main' into upgrade/react-native-webview

jpcloureiro committed:
Merge branch 'main' into upgrade/react-native-webview

jpcloureiro committed:
bring quotes snapshot from main

jpcloureiro committed:
update jest Quotes snapshot

jpcloureiro committed:
Merge branch 'main' into upgrade/react-native-webview

legobeat committed:
Merge branch 'main' into upgrade/react-native-webview

jpcloureiro committed:
Merge branch 'main' into upgrade/react-native-webview

jpcloureiro committed:
update PPOMView jest snapshot

jpcloureiro committed:
RNWebView: (inpage provider) Replace script tag with straight code execution (#10053)

Update `inpage-bridge` build process and change provider injection method

Context:

CSP directives might block inline script execution, thus preventing our
web3 provider from being injected with a script tag creation.

Replacing it with a direct function call avoids being blocked. 


Co-authored-by: jiexi <jiexiluan@gmail.com>

jpcloureiro committed:
Merge branch 'main' into upgrade/react-native-webview

jpcloureiro commented:
PR ready for review.

Context on `inpage-bridge` changes [here](https://github.com/MetaMask/metamask-mobile/pull/10053)

jpcloureiro committed:
remove unneeded trycatch block

jpcloureiro committed:
tests: (Jest) ignore inpage bridge file

InpageBridgeWeb3.js has changed and now contains only runnable code,
which is now considered as untested code by Jest.

Before, most of the code was wrapped in single quotes.

jpcloureiro committed:
Merge branch 'main' into upgrade/react-native-webview

tommasini reviewed (dismissed):
LGTM!

jpcloureiro commented:
[E2E Regression Tests](https://app.bitrise.io/app/be69d4368ee7e86d/pipelines/5f7e59bd-f012-4e65-8df7-821b9b752fe5): Pass ✅

jpcloureiro committed:
Merge branch 'main' into upgrade/react-native-webview

jpcloureiro commented:
[E2E Regression Tests](https://app.bitrise.io/app/be69d4368ee7e86d/pipelines/cbc87adb-b113-44f3-8de4-fbe426ced83f) on 056cf00  pass ✅

legobeat committed:
Merge branch 'main' into upgrade/react-native-webview

FrederikBolding reviewed (approved):
Snaps changes LGTM and confirmed working 👍
